share/models/fields.py

```diff
@@ -97,66 +105,137 @@ def validate(self, value, model_instance):
 # instances of their base class that contribute instead. Keep track of these
 # fields that do a Share field's job, so they'll show up in _meta.get_fields().
 class ShareRelatedField:
+
+    PRETENDING_TO_BE = None
+
     def __init__(self, *args, **kwargs):
+        # Correct M2M fields
+        if 'to' in kwargs and not args:
+            args = args + (kwargs.pop('to'), )
+
+        self._kwargs = kwargs
         self.__equivalent_fields = set()
         super().__init__(*args, **kwargs)
 
     def add_equivalent_fields(self, *fields):
         self.__equivalent_fields.update(fields)
 
-    def __eq__(self, other):
-        if other in self.__equivalent_fields:
-            return True
-        return super().__eq__(other)
-
-    def __hash__(self):
-        return super().__hash__()
-
-
-class ShareOneToOneField(ShareRelatedField, models.OneToOneField):
-    def __init__(self, model, **kwargs):
-        self._kwargs = kwargs
-        super().__init__(model, **kwargs)
-
     def contribute_to_class(self, cls, name, **kwargs):
-        actual = self.__class__.mro()[2](self.remote_field.model, **self._kwargs)
+        actual = self.PRETENDING_TO_BE(self.remote_field.model, **self._get_kwargs(cls))
         actual.contribute_to_class(cls, name, **kwargs)
 
         if isinstance(self.remote_field.model, str):
-            version = self.__class__.mro()[2](self.remote_field.model + 'Version', editable=False, **{**self._kwargs, 'db_index': False})
+            version_model = self.remote_field.model + 'Version'
```

Comment:
Would it be worth checking if the named model is a ShareObject subclass, and return if not?